sql, server: fix errors seen from combined statements endpoint

xinhaoz opened the pull request:
## Commit 1 sql, server: use a view to join persisted and in-mem sql stats
Partially addresses: #71245

Previously, we created virtual tables to join in-memory and
persisted disk statement and transaction statistics. This
proved to be inefficient as requests to the the virtual
tables lead to full scans of the underlying system tables.

This commit utilizes virtual views to combine in-memory and disk
stats. The view queries in-memory stats from the virtual tables
`crdb_internal.cluster_{statement, transaction}_statistics,`
and combines them with the results from the system tables.
This allows us to push down query filters into the system
tables, leveraging their existing indexes.

Release note: None

## Commit 2 sql: decrease stats flush interval to every 10 mins 

Previously, we flush in-memory sql stats collected by each
node on an hourly interval. We have found that this hourly
interval might be too conservative, and the size of the
returned cluster-wide stats after an hour can also be
quite large, sometimes exceeding the gRPC max message
size.

This commit lowers the flush interval to every 10 minutes.
Since we want to continue to aggregate stats on an hourly
interval, we introduce a new cluster setting
`sql.stats.aggregation.interval` to control the
aggregation interval separately from the flush frequency.

Release note (sql change): The default sql stats flush interval
is now 10 minutes. A new cluster setting
`sql.stats.aggregation.interval` controls the aggregation
interval of sql stats, with a default value of 1 hour.

## Commit 3 server: allow statements EP to optionally exclude stmt or txns stats

Closes: cockroachdb#71829

Previously, the /statements endpoint returned cluster-wide
in-memory stats, containing  both statements and transactions stats.
In the past, we've observed the Statements endpoint response being
too large for gRPC. Because this endpoint is being used by virtual
tables that powers our combined stats api,
cluster_{statement,transactions}_stats, we might continue to surpass
the gRPC message size in the new api. However, each virtual table
only uses roughly half the response size (either statement or txn
stats).

This commit allows the virtual tables to exclude statement or txn
stats from the Statements endpoint resposne by introducing new
request parameters to /statements. This reduces the response size
in the stats virtual tables.

Release note: None

xinhaoz committed:
sql, server: use a view to join persisted and in-mem sql stats

Partially addresses: #71245

Previously, we created virtual tables to join in-memory and
persisted disk statement and transaction statistics. This
proved to be inefficient as requests to the the virtual
tables lead to full scans of the underlying system tables.

This commit utilizes virtual views to combine in-memory and disk
stats. The view queries in-memory stats from the virtual tables
`crdb_internal.cluster_{statement, transactions}_statistics`,
and combines them with the results from the system tables.
This allows us to push down query filters into the system
tables, leveraging their existing indexes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74831)
<!-- Reviewable:end -->

xinhaoz committed:
sql: decrease stats flush interval to every 10 mins

Closes: #71245

Previously, we flush in-memory sql stats collected by each
node on an hourly interval. We have found that this hourly
interval might be too conservative, and the size of the
returned cluster-wide stats after an hour can also be
quite large, sometimes exceeding the gRPC max message
size.

This commit lowers the flush interval to every 10 minutes.
Since we want to continue to aggregate stats on an hourly
interval, we introduce a new cluster setting
`sql.stats.aggregation.interval` to control the
aggregation interval separately from the flush frequency.

Release note (sql change): The default sql stats flush interval
is now 10 minutes. A new cluster setting
`sql.stats.aggregatinon.interval` controls the aggregation
interval of sql stats, with a default value of 1 hour.

xinhaoz committed:
server: allow statements EP to optionally exclude stmt or txns stats

Closes: #71829

Previously, the /statements endpoint returned cluster-wide
in-memory stats, containing  both statements and transactions stats.
In the past, we've observed the Statements endpoint response being
too large for gRPC. Because this endpoint is being used by virtual
tables that powers our combined stats api,
cluster_{statement,transactions}_stats, we might continue to surpass
the gRPC message size in the new api. However, each virtual table
only uses roughly half the response size (either statement or txn
stats).

This commit allows the virtual tables to exclude statement or txn
stats from the Statements endpoint resposne by introducing new
request parameters to /statements. This reduces the response size
in the stats virtual tables.

Release note: None

xinhaoz commented:
Not sure if we want to include the third commit in this fix as it wasn't discussed too much, so I can remove it or create a separate PR for it.

ajwerner commented:
> Previously, the /statements endpoint returned cluster-wide
in-memory stats, containing both statements and transactions stats.
In the past, we've observed the Statements endpoint response being
too large for gRPC.

Was the response to the browser endpoint where we hit the size issues or the response to the internal gRPC call to collect stats from each node?

maryliag reviewed (commented):
> Reviewed 15 of 21 files at r1, 4 of 6 files at r2, 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @xinhaoz)

___
*[pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go, line 85 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtP62s67zdPipofyYnu:-MtP62s67zdPipofyYnv:boy0zxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go#L85)):*
> ```Go
> // SQLStatsAggregationInterval is the cluster setting that controls the aggregation
> // interval for stats when we flush to disk
> var SQLStatsAggregationInterval = settings.RegisterDurationSetting(
> ```

nit: should we add some comment or check that this value is always bigger than `sql.stats.flush.interval`?
___
*[pkg/cli/zip_test.go, line 93 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtP769xEPbFf26SyGls:-MtP769xEPbFf26SyGlt:bu6dsbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/cli/zip_test.go#L93)):*
> ```Go
> 	'statement_statistics',
> 	'transaction_statistics',
> 	'cluster_transaction_statistics',
> ```

you added this twice


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
I tried this out and it works nicely. I think it strikes a good balance. Do we have any clusters with high traffic and large node counts to validate this on? Once upon a time there were long running clusters where such bugs would have been obvious. I hope we can bring them back one day. Curious about the methodology we have today.

I think it'd be nice to see some `execbuilder` logic tests that show query plans for queries made by the various requests which demonstrate that they don't end up doing any full table scans. You can add a new file or two to `sql/opt/exec/execbuilder/testdata` which have the queries as they'd be issued from the http endpoint and have it rewrite the plans. That'll provide confidence and prevent regressions. 

When I looked at the queries I was a little surprised to see no limits. I wonder if that may prove to also be a key part of the problem here. 

---

The thing that leaves me most anxious is a race condition where either we see the stats in-memory and then we also see them on disk or we don't see them in either place because they've just been flushed and our out of memory but our disk read was at a timestamp before the flush. Solving that problem definitely shouldn't happen here, but it's something we should track. Solving it is a little tricky. We'd to make sure to ship a timestamp in the RPCs and we'd need to hold on to some in-memory state after we've flushed it to serve. The follower reads make this problem much much more common. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-:-MtP3VmT6O-dTeIbIxMj:b-o2c6od)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @maryliag, and @xinhaoz)

___
*[pkg/sql/crdb_internal.go, line 4986 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPARx07yyT5HQaaKr4:-MtPARx07yyT5HQaaKr5:b-7blifp) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/crdb_internal.go#L4986)):*
> ```Go
> }
> 
> var crdbInternalClusterStmtStatsTable = virtualSchemaTable{
> ```

I think this deserves commentary noting that it will only contain data that has not yet been flushed to the persistent system tables. Same for the other new table. 
___
*[pkg/sql/crdb_internal.go, line 5086 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPAtbZ8y8Zx4JJQ_3A:-MtPAtbZ8y8Zx4JJQ_3B:bu07mbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/crdb_internal.go#L5086)):*
> ```Go
> }
> 
> var crdbInternalStmtStatsView = virtualSchemaView{
> ```

This deserves commentary explaining what it is and why it exists. Same for the other view. 
___
*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 646 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtP2_L246GGYQopHFQd:-MtP2_L3Aie0tAMRei5D:bzagahe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c8de21f168562290f59ee6fa4254f949da67c9d/pkg/sql/logictest/testdata/logic_test/crdb_internal#L646)):*
> ```
> /Table/62/2/1  /Table/62/2/2
> /Table/62/2/2  /Table/62/2/3
> /Table/62/2/3  /Table/62/3/1
> ```

I think you can lose this. I suspect it was added by the rewrite. 



_Code quote:_
```
/Table/62/1/1  /Table/62/1/2
/Table/62/1/2  /Table/62/1/3
/Table/62/1/3  /Table/62/2/1
/Table/62/2/1  /Table/62/2/2
/Table/62/2/2  /Table/62/2/3
/Table/62/2/3  /Table/62/3/1
```


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @maryliag, and @xinhaoz)

___
*[pkg/server/combined_statement_stats.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPQK6J2BgNfFYnOoJC:-MtPQK6J2BgNfFYnOoJD:bqeeesl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c8de21f168562290f59ee6fa4254f949da67c9d/pkg/server/combined_statement_stats.go#L109)):*
> ```Go
> 	}
> 
> 	followerReadClause := "AS OF SYSTEM TIME follower_read_timestamp()"
> ```

quick refactor suggestion: since the pattern of checking the testing knobs and return the string representation of the AOST clause comes up so frequently, we can refactor this logic into a method defined on the `sqlstats.TestingKnobs`.

``` golang
func (t *TestingKnobs ) GetAOSTClause() string {
	if testingKnobs != nil {
		return testingKnobs.AOSTClause
	}
       
	return "AS OF SYSTEM TIME follower_read_timestamp()"
}
```
___
*[pkg/server/tenant_status.go, line 466 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPRa8KCsgRow7dQ8xl:-MtPRa8KCsgRow7dQ8xm:buhgupp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c8de21f168562290f59ee6fa4254f949da67c9d/pkg/server/tenant_status.go#L466)):*
> ```Go
> 
> 	return getCombinedStatementStats(ctx, req, t.sqlServer.pgServer.SQLServer.GetSQLStatsProvider(),
> 		t.sqlServer.internalExecutor, t.sqlServer.execCfg.SQLStatsTestingKnobs)
> ```

minor style nit: this function call is getting quite long, it would improve the readability a bit by vertically align the arguments 

``` golang
	return getCombinedStatementStats(
                        ctx, 
                        req, 
                        t.sqlServer.pgServer.SQLServer.GetSQLStatsProvider(),
		        t.sqlServer.internalExecutor, 
                        t.sqlServer.execCfg.SQLStatsTestingKnobs,
                   )

```
___
*[pkg/server/serverpb/status.proto, line 1204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPTGnc18qYGSkXbWqT:-MtPTGnc18qYGSkXbWqU:b-u2bwe5) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/server/serverpb/status.proto#L1204)):*
> ```protobuf
>   int64 end = 4 [(gogoproto.nullable) = true];
>   bool exclude_statements = 5;
>   bool exclude_transactions = 6;
> ```

Don't think I'm a very big fan of bolting on boolean flags onto an already big endpoint. Is there any reason this method is in favour over new split endpoint approach? 

If we really need to go with this route, I would prefer an "inclusive" flag over "exclusive" flag. Or even better, perhaps defining an enum here:

```proto
enum FetchMode {
  StmtAndTxnStats = 0;
  StmtStatsOnly = 1;
  TxnStatsOnly = 2;
}
```
___
*[pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go, line 85 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtP62s67zdPipofyYnu:-MtPSZGICBZtmfCUE0_0:b-w66aav) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go#L85)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

nit: should we add some comment or check that this value is always bigger than `sql.stats.flush.interval`?
</blockquote></details>

Hmm it's a bit difficult to write validation function directly for this cluster setting. The closest thing we have is probably `SetOnChange` callback that we can register to validate the input, and perhaps logs a warning message if we detect any anomaly. 


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @maryliag, and @xinhaoz)

___
*[pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go, line 85 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtP62s67zdPipofyYnu:-MtPVDyN0o-Y30E_4CjZ:b-5l3diw) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go#L85)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm it's a bit difficult to write validation function directly for this cluster setting. The closest thing we have is probably `SetOnChange` callback that we can register to validate the input, and perhaps logs a warning message if we detect any anomaly. 
</blockquote></details>

My concern was that if that value is smaller, if that could cause any issues. 
One solution could be here that here (https://github.com/cockroachdb/cockroach/pull/74831/files#diff-070ff15c495b929697bdcf0d193a22d0c5e1cbafa596cf6dbf4731bbbf8e7144R255) we compare the values and if the aggregation is smaller, return the flush value instead?


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @maryliag, and @xinhaoz)

___
*[pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go, line 85 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtP62s67zdPipofyYnu:-MtPWX5PAv93MWAVvxeH:b-4fqcy9) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go#L85)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

My concern was that if that value is smaller, if that could cause any issues. 
One solution could be here that here (https://github.com/cockroachdb/cockroach/pull/74831/files#diff-070ff15c495b929697bdcf0d193a22d0c5e1cbafa596cf6dbf4731bbbf8e7144R255) we compare the values and if the aggregation is smaller, return the flush value instead?
</blockquote></details>

Hmm good point. We could potentially override that if the cluster setting doesn't make sense at all. Perhaps we go one step further and just directly take this into our own hand? If we see the agg interval setting doesn't make sense, we just go ahead and set it to flush interval. Though we should probably:
1. log a message saying that we are overriding user's setting
2. be explicit on the release note that we will have this behavior.


<!-- Sent from Reviewable.io -->

xinhaoz commented:
___
*[pkg/server/serverpb/status.proto, line 1204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPTGnc18qYGSkXbWqT:-Mti32t63HGKFeBFADrO:b-umbu17) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/server/serverpb/status.proto#L1204)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Don't think I'm a very big fan of bolting on boolean flags onto an already big endpoint. Is there any reason this method is in favour over new split endpoint approach? 

If we really need to go with this route, I would prefer an "inclusive" flag over "exclusive" flag. Or even better, perhaps defining an enum here:

```proto
enum FetchMode {
  StmtAndTxnStats = 0;
  StmtStatsOnly = 1;
  TxnStatsOnly = 2;
}
```
</blockquote></details>

I threw this in to help mitigate the response size issues in a way that wouldn't require any changes to the frontend (and potentially further considerations or modifications). We can still go with a split endpoint for stmts and txns in the future, or if we'd rather not expand this EP more and wait for the former to be done, I can remove this commit from the PR. 


<!-- Sent from Reviewable.io -->

jocrl reviewed (commented):
nit: For the note in commit 2,
> Release note (sql change): The default sql stats flush interval
is now 20 minutes...

I think this should be 10 minutes, not 20?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-:-MtiEvKY7Id5DLyBH5iv:b-fcbqza)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @maryliag, and @xinhaoz)



<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @maryliag, and @xinhaoz)

___
*[pkg/server/serverpb/status.proto, line 1204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPTGnc18qYGSkXbWqT:-MtilSUn4MpAnlrrItMa:b8x7bo6) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/server/serverpb/status.proto#L1204)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

I threw this in to help mitigate the response size issues in a way that wouldn't require any changes to the frontend (and potentially further considerations or modifications). We can still go with a split endpoint for stmts and txns in the future, or if we'd rather not expand this EP more and wait for the former to be done, I can remove this commit from the PR. 
</blockquote></details>

Let's split the last commit into its own PR since it's less so urgent for 21.2.5 release. We can iterate on the API design a bit more. 


<!-- Sent from Reviewable.io -->

xinhaoz commented:
___
*[pkg/server/serverpb/status.proto, line 1204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPTGnc18qYGSkXbWqT:-MtjPk-T7u-kjkujel-V:bwttyid) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/server/serverpb/status.proto#L1204)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Let's split the last commit into its own PR since it's less so urgent for 21.2.5 release. We can iterate on the API design a bit more. 
</blockquote></details>

Are we sure that if we don't include this, we won't still be in danger of message size being too large? I can change to the enum format, but I'm not convinced we don't want to include it for 21.2.5 if this will help reduce errors @maryliag 


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @maryliag, and @xinhaoz)

___
*[pkg/server/serverpb/status.proto, line 1204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPTGnc18qYGSkXbWqT:-MtjRUrW06cmjMtAcwkJ:b-2j3k79) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/server/serverpb/status.proto#L1204)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Are we sure that if we don't include this, we won't still be in danger of message size being too large? I can change to the enum format, but I'm not convinced we don't want to include it for 21.2.5 if this will help reduce errors @maryliag 
</blockquote></details>

It's a small change with low risk and a possible great improvement, we should keep this commit here and also include on 21.2.5


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @maryliag, and @xinhaoz)

___
*[pkg/server/serverpb/status.proto, line 1204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPTGnc18qYGSkXbWqT:-MtjSHCp3ccaEGyu0UxQ:b-edt4tz) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/server/serverpb/status.proto#L1204)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

It's a small change with low risk and a possible great improvement, we should keep this commit here and also include on 21.2.5
</blockquote></details>

My reasoning was that, this change definitely benefits reducing the payload size between backend RPC that collects in-memory cluster-wide sql stats, but eventually, when we send that data to the frontend, we will still include all the data.

Anyway, not strong preference here but I would love to see it as an enum rather than two boolean flags. 


<!-- Sent from Reviewable.io -->

xinhaoz commented:
___
*[pkg/server/serverpb/status.proto, line 1204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPTGnc18qYGSkXbWqT:-MtjYH72-p73kZ3uTGA9:b-80xfce) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/server/serverpb/status.proto#L1204)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

My reasoning was that, this change definitely benefits reducing the payload size between backend RPC that collects in-memory cluster-wide sql stats, but eventually, when we send that data to the frontend, we will still include all the data.

Anyway, not strong preference here but I would love to see it as an enum rather than two boolean flags. 
</blockquote></details>

Yeah that's true true, my concern was that even after lowering the flush interval to 10 minutes we could potentially run into the message size error in the virtual table calls to Statement. Sounds good, I'll change to an enum and I'll also file a new ticket for splitting the API completely :thumbsup: 


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @maryliag, and @xinhaoz)

___
*[pkg/server/combined_statement_stats.go, line 115 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtnXYU-8TvIsytpi5nJ:-MtnXYU-8TvIsytpi5nK:b712jdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cd1821391be01c0e082de67e6f97f53f5062a00/pkg/server/combined_statement_stats.go#L115)):*
> ```Go
> 	followerReadClause := testingKnobs.GetAOSTClause()
> 
> 	return fmt.Sprintf("%s %s", followerReadClause, buffer.String()), args
> ```

nit: let's just directly write `followerReadClause` into the `buffer` above instead of using the `Sprintf` here ? 


<!-- Sent from Reviewable.io -->

xinhaoz commented:
___
*[pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go, line 85 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtP62s67zdPipofyYnu:-MtsLeAf4M_8hbOCWl52:b-jgye1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go#L85)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm good point. We could potentially override that if the cluster setting doesn't make sense at all. Perhaps we go one step further and just directly take this into our own hand? If we see the agg interval setting doesn't make sense, we just go ahead and set it to flush interval. Though we should probably:
1. log a message saying that we are overriding user's setting
2. be explicit on the release note that we will have this behavior.
</blockquote></details>

We'll have to enforce the same thing with the flush cluster setting. I noticed we don't explicitly set any cluster settings (aside from overrides in tests. Is this behaviour something we're okay with changing for this purpose, or would we rather just log a message that the invalid cluster setting will be ignored, and take the max(agg, flush) for the agg interval, and min(agg, flush) for the flush interval? We could also create something like `SetOnChange` that would run before the cluster setting is changed and return an error so it's never changed in the first place.


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @kevin-v-ngo, @maryliag, and @xinhaoz)

___
*[pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go, line 85 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtP62s67zdPipofyYnu:-MtsR8v-33z31FYyXcqI:bsslhrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go#L85)):*
> We could also create something like SetOnChange that would run before the cluster setting is changed

Unfortunately, `SetOnChange` is [executed](https://github.com/cockroachdb/cockroach/blob/b08695e9160a479e265eee5e7631df76107484eb/pkg/settings/values.go#L152) only after the setting is changed, we can't really block the setting from changing using the callback passed in `SetOnChange`. 

I can think of 2 options here:

1. we can alternatively mark the agg_interval cluster setting private and only allow the user to set it through our builtins. E.g. `SELECT crdb_internal.set_sql_stats_agg_interval(...)` and within that builtin we can perform the validation check. 
2. we allow user to change the cluster setting however they want, but in the docs we just be explicit about the `max(agg, flush)` behavior. 

cc: @kevin-v-ngo   thoughts? 


<!-- Sent from Reviewable.io -->

kevin-v-ngo commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @kevin-v-ngo, @maryliag, and @xinhaoz)
> 
> _[pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go, line 85 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtP62s67zdPipofyYnu:-MtsR8v-33z31FYyXcqI:bsslhrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go#L85)):_
> 
> > We could also create something like SetOnChange that would run before the cluster setting is changed
> 
> Unfortunately, `SetOnChange` is [executed](https://github.com/cockroachdb/cockroach/blob/b08695e9160a479e265eee5e7631df76107484eb/pkg/settings/values.go#L152) only after the setting is changed, we can't really block the setting from changing using the callback passed in `SetOnChange`.
> 
> I can think of 2 options here:
> 
> 1. we can alternatively mark the agg_interval cluster setting private and only allow the user to set it through our builtins. E.g. `SELECT crdb_internal.set_sql_stats_agg_interval(...)` and within that builtin we can perform the validation check.
> 2. we allow user to change the cluster setting however they want, but in the docs we just be explicit about the `max(agg, flush)` behavior.
> 
> cc: @kevin-v-ngo thoughts?

I don't think we need introduce a built-in. I was chatting with @xinhaoz offline on this issue. We currently do not recommend users to change the aggregation interval (it's by default 1 hour) where the same guidance should be provided for the flush interval. We'll make sure the documentation highlights that these two settings should not be configured especially with agg greater than the flush interval.

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @kevin-v-ngo, @maryliag, and @xinhaoz)

___
*[pkg/server/combined_statement_stats.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPQK6J2BgNfFYnOoJC:-MtjN_kO59OWGd9sNxs9:b-j91rae) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c8de21f168562290f59ee6fa4254f949da67c9d/pkg/server/combined_statement_stats.go#L109)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

quick refactor suggestion: since the pattern of checking the testing knobs and return the string representation of the AOST clause comes up so frequently, we can refactor this logic into a method defined on the `sqlstats.TestingKnobs`.

``` golang
func (t *TestingKnobs ) GetAOSTClause() string {
	if testingKnobs != nil {
		return testingKnobs.AOSTClause
	}
       
	return "AS OF SYSTEM TIME follower_read_timestamp()"
}
```
</blockquote></details>

Nice recDone.
___
*[pkg/server/tenant_status.go, line 466 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPRa8KCsgRow7dQ8xl:-MtjP_bW8RQZvbrK7dMH:b-lkx7oc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c8de21f168562290f59ee6fa4254f949da67c9d/pkg/server/tenant_status.go#L466)):*
> if testingKnobs != nil {
> 		return testingKnobs.AOSTClause
> 	}
>        
> 	return "AS OF SYSTEM TIME follower_read_timestamp()"
> }

Done.
___
*[pkg/sql/crdb_internal.go, line 4986 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPARx07yyT5HQaaKr4:-MtjRbZB-tPDMDduYb_d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/crdb_internal.go#L4986)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think this deserves commentary noting that it will only contain data that has not yet been flushed to the persistent system tables. Same for the other new table. 
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 5086 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPAtbZ8y8Zx4JJQ_3A:-MtjVOxu5YC9rVZ2APvp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/crdb_internal.go#L5086)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This deserves commentary explaining what it is and why it exists. Same for the other view. 
</blockquote></details>

Done.
___
*[pkg/cli/zip_test.go, line 93 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtP769xEPbFf26SyGls:-Mti51sK9agGF1Bi9I9h:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/cli/zip_test.go#L93)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

you added this twice
</blockquote></details>

Removed.


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
Code mostly looks good, just some very small nits. Will stamp it once CI is green

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-:-MuBtppyE88dYb3OYPqK:b-7v8kae)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @kevin-v-ngo, @maryliag, and @xinhaoz)

___
*[pkg/server/combined_statement_stats.go, line 139 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuBrbi8DuBBpuUsw8SX:-MuBrbi8DuBBpuUsw8SY:b-lrv0zr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae69acc247e26790c91f12d3510a9278d9f1e39/pkg/server/combined_statement_stats.go#L139)):*
> ```Go
> 				app_name,
> 				aggregated_ts,
> 				metadata,
> ```

Let's open an issue to track a new endpoint that prettify the statement just for the Statement Detail Page.
___
*[pkg/sql/crdb_internal.go, line 4986 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPARx07yyT5HQaaKr4:-MuBsaYC-kSQk50oXmp2:b-lx4aqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/crdb_internal.go#L4986)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think the comment here is still missing.
___
*[pkg/sql/crdb_internal.go, line 5086 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtPAtbZ8y8Zx4JJQ_3A:-MuBsm8I7hE7cpUbevZh:bfqp5m3) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/crdb_internal.go#L5086)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Don't forget to add the similar comments for the txn view.
___
*[pkg/sql/crdb_internal.go, line 4989 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuBs6-VEt9CxIyouhGR:-MuBs6-VEt9CxIyouhGS:b-4k2p2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae69acc247e26790c91f12d3510a9278d9f1e39/pkg/sql/crdb_internal.go#L4989)):*
> ```Go
> var crdbInternalClusterStmtStatsTable = virtualSchemaTable{
> 	comment: `cluster-wide statement statistics that have not yet been flushed ` +
> 		`to system tables. Querying this table is a somewhat expensive operation ` +
> ```

small preference nit: `s/is a somewhat expensive/is an expensive/`. (Since we are doing RPC fanout unconditionally) 
___
*[pkg/sql/crdb_internal.go, line 5104 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuBtNpk8CbLokmFtih9:-MuBtNpk8CbLokmFtihA:bjmuzr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae69acc247e26790c91f12d3510a9278d9f1e39/pkg/sql/crdb_internal.go#L5104)):*
> ```Go
> 	max(metadata) as metadata,
> 	crdb_internal.merge_statement_stats(array_agg(statistics)),
>   max(sampled_plan),
> ```

super small nit: the indentation is not aligned
___
*[pkg/sql/crdb_internal.go, line 5301 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuBtRYn2kLH7BQuA_J1:-MuBtRYn2kLH7BQuA_J2:b-7pxrbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae69acc247e26790c91f12d3510a9278d9f1e39/pkg/sql/crdb_internal.go#L5301)):*
> ```Go
> 	max(metadata),
> 	crdb_internal.merge_transaction_stats(array_agg(statistics)),
>   aggregation_interval
> ```

super small nits here on indentation.
___
*[pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go, line 85 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MtP62s67zdPipofyYnu:-MuBqOSyBeQr1q5MWBaj:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/429aca900e2e72dee20d643c31f2b7857ce58124/pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go#L85)):*
<details><summary><i>Previously, kevin-v-ngo wrote…</i></summary><blockquote>

I don't think we need introduce a built-in. I was chatting with @xinhaoz offline on this issue. We currently do not recommend users to change the aggregation interval (it's by default 1 hour) where the same guidance should be provided for the flush interval. We'll make sure the documentation highlights that these two settings should not be configured especially with agg greater than the flush interval.
</blockquote></details>

:+1:


<!-- Sent from Reviewable.io -->

ajwerner commented on pkg/sql/crdb_internal.go:
```diff
@@ -5087,6 +5086,70 @@ CREATE TABLE crdb_internal.statement_statistics (
+	FROM
+			crdb_internal.cluster_statement_statistics
+	UNION ALL
+			SELECT
+					aggregated_ts,
+					fingerprint_id,
+					transaction_fingerprint_id,
+					plan_hash,
+					app_name,
+					metadata,
+					statistics,
+					plan,
+					agg_interval
+			FROM
+					system.statement_statistics
```
the suggestion is that we do a group by and aggregate here as well as above to avoid needing to read everything, I'm not sure if the limit pushdown will work perfectly, but hopefully.

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @kevin-v-ngo, @maryliag, and @xinhaoz)

___
*[pkg/server/combined_statement_stats.go, line 140 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuD8hx7AhOleKQ56x4L:-MuD8hx7AhOleKQ56x4M:bopvua0) ([raw file](https://github.com/cockroachdb/cockroach/blob/38c69e59c05c992e471ef52a1ef32fc14c42a2cc/pkg/server/combined_statement_stats.go#L140)):*
> ```Go
>         app_name,
>         aggregated_ts,
>         jsonb_set(
> ```

I think you might want to rebase against @thardy98 's PR or there's gonna be merge conflict.
___
*[pkg/server/combined_statement_stats.go, line 242 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuD9-Jy1gpWdBFOzz0Y:-MuD9-Jy1gpWdBFOzz0Z:b-xjvtt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/38c69e59c05c992e471ef52a1ef32fc14c42a2cc/pkg/server/combined_statement_stats.go#L242)):*
> ```Go
> 
> func collectCombinedTransactions(
> 	ctx context.Context, ie *sql.InternalExecutor, whereClause *string, qargs []interface{},
> ```

Any particular reason it's a string pointer here? In Golang string is just a struct with length and pointer to underlying bytes, so the overhead of passing the string is not that high.


<!-- Sent from Reviewable.io -->

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @kevin-v-ngo, @maryliag, and @xinhaoz)

___
*[pkg/server/combined_statement_stats.go, line 139 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuBrbi8DuBBpuUsw8SX:-MuDEyD35zIA97IJQhvX:b2bsujc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae69acc247e26790c91f12d3510a9278d9f1e39/pkg/server/combined_statement_stats.go#L139)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Let's open an issue to track a new endpoint that prettify the statement just for the Statement Detail Page.
</blockquote></details>

Ah right, rebased and filed: https://github.com/cockroachdb/cockroach/issues/75466
___
*[pkg/server/combined_statement_stats.go, line 242 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuD9-Jy1gpWdBFOzz0Y:-MuDBDAX-uZdCnElvp4A:b-dmeitn) ([raw file](https://github.com/cockroachdb/cockroach/blob/38c69e59c05c992e471ef52a1ef32fc14c42a2cc/pkg/server/combined_statement_stats.go#L242)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Any particular reason it's a string pointer here? In Golang string is just a struct with length and pointer to underlying bytes, so the overhead of passing the string is not that high.
</blockquote></details>

TIL! Removed the pointer for the clause string.


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
Hmm, seems like CI is complaining about malformed SQL statement

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-:-MuDI9hI2Pe0Foc9_7d9:blp5egn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @kevin-v-ngo, @maryliag, and @xinhaoz)

___
*[pkg/sql/crdb_internal.go, line 5305 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuDI0YJ1ZocmqrNUUvJ:-MuDI0YK6LQNsq74nM3Q:b-h3953r) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d5272532beb26b9f40f2e23305a0b18170fff0f/pkg/sql/crdb_internal.go#L5305)):*
> ```Go
> 	max(metadata),
> 	crdb_internal.merge_transaction_stats(array_agg(statistics)),
>   aggregation_interval
> ```

small nit: misaligned indentation 


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
This is looking great! Added a few nits and comments

> Reviewed 7 of 42 files at r10, 29 of 35 files at r11, 2 of 2 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-:-MuDjioFE0DDEhcQQ_07:btmhfeu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @kevin-v-ngo, @maryliag, and @xinhaoz)

___
*[pkg/sql/sqlstats/test_utils.go, line 38 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuDiTsD7TRI3owUESE2:-MuDiTsD7TRI3owUESE3:bxf6hfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/617ed5e885fb2120c90bcd7c7bf43bbd94de375a/pkg/sql/sqlstats/test_utils.go#L38)):*
> ```Go
> 
> // GetAOSTClause returns the appropriate AS OF SYSTEM TIME clause to be
> //used when reading from statements and transactions system tables.
> ```

small nit: space after // 
___
*[pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go, line 85 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuDimh6AANxU8vd7YS8:-MuDimh6AANxU8vd7YS9:bjq19qa) ([raw file](https://github.com/cockroachdb/cockroach/blob/617ed5e885fb2120c90bcd7c7bf43bbd94de375a/pkg/sql/sqlstats/persistedsqlstats/cluster_settings.go#L85)):*
> ```Go
> 
> // SQLStatsAggregationInterval is the cluster setting that controls the aggregation
> // interval for stats when we flush to disk
> ```

small nit: period
___
*[pkg/ui/workspaces/cluster-ui/src/statementsPage/statementsPage.tsx, line 285 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuDj0dY6pQVUQw7jDPc:-MuDj0dY6pQVUQw7jDPd:b-hap02p) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e6c3619cee958722b745a56f8a59f9baa2e37d1/pkg/ui/workspaces/cluster-ui/src/statementsPage/statementsPage.tsx#L285)):*
> ```XML
> 
>   componentDidUpdate = (): void => {
>     this.updateQueryParams();
> ```

if you remove the updateQueryParams from here, it won't properly update the query params when a tab is changed under the SQL Activity page. That is the main reason we need to keep it here


<!-- Sent from Reviewable.io -->

xinhaoz commented:
___
*[pkg/ui/workspaces/cluster-ui/src/statementsPage/statementsPage.tsx, line 285 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-MuDj0dY6pQVUQw7jDPc:-MuEEwNk8FU2wWwmdgo4:b-t9b28z) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e6c3619cee958722b745a56f8a59f9baa2e37d1/pkg/ui/workspaces/cluster-ui/src/statementsPage/statementsPage.tsx#L285)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

if you remove the updateQueryParams from here, it won't properly update the query params when a tab is changed under the SQL Activity page. That is the main reason we need to keep it here
</blockquote></details>

Oh these changes weren't meant to be included here, so I'll remove them from this PR.


<!-- Sent from Reviewable.io -->

Azhng reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :rocket: :fire: Great work here! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-:-MuEOxcK-50ziWaFJoEG:brpa5cq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @kevin-v-ngo, @maryliag, and @xinhaoz)



<!-- Sent from Reviewable.io -->

maryliag reviewed (approved):
Awesome work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :rocket: 

> Reviewed 1 of 35 files at r11, 3 of 3 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74831#-:-MuGLuiC4oWyiEqafIyi:b-52jp9q)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner and @xinhaoz)



<!-- Sent from Reviewable.io -->

xinhaoz commented:
TFTR!!!
bors r=Azhng,maryliag
